Add workspace-vscode-python image

Victorsss-Orz opened the pull request:
Developed by @firasm and @echuber2 , docker image that supports vscode in workspace and running python with vscode.

nwalters512 commented on workspaces/vscode-python/settings.json:
```diff
@@ -0,0 +1,56 @@
+    // "editor.codeActionsOnSave": {
+
+    // },
+}
```
You're missing a trailing newline.

nwalters512 commented on workspaces/vscode-python/settings.json:
Many of the settings here seem opinionated (e.g. `workbench.colorTheme`, `files.exclude`, `problems.visibility`, `editor.lightbulb.enabled`). I agree with disabling telemetry, but this base image should be as unopinionated as possible.

echuber2 replied:
I would vote in favor of setting a dark theme by default if that's what people would expect for VS Code anyway. For a while, code-server was popping up a nag message to encourage trying out the dark modern theme, so I was setting _that_ by default to remove the nag message, if nothing else.

nwalters512 replied:
I think setting a dark theme by default would be reasonable.

nwalters512 commented on workspaces/vscode-python/settingsEmpty.json:
Why is this empty file here?

firasm replied:
The empty files are there because this particular setup **restricts** students from being able to make changes to the settings.

I'm not sure how PL wants to handle this - in my course, it was decided to disable the ability of users to customize their editor and there's no way to allow certain things to be edited and not others.

Perhaps we don't want that more generally? Though, I still think some of these configurations should be left in the Dockerfile "template" so others could customize as needed...

nwalters512 replied:
This sounds like the kind of opinionated choice that doesn't belong in an official image.

What's being contributed here is not a "template"; it's a full-blown image that will be automatically built and published as `prairielearn/workspace-vscode-python`. If you'd like to _document_ in the workspace's `README.md` how one can build an image on top of `prairielearn/workspace-vscode-python`, that's fine. More generally, we fully expect that instructors may need to build customized images on top of ours if they have specific requirements.

nwalters512 commented on workspaces/vscode-python/workspaceTemplate/.vscode/settings.json:
Why is this empty file here?

nwalters512 reviewed (commented):
You'll need to add this image to our CI workflows:

- https://github.com/PrairieLearn/PrairieLearn/blob/master/.github/workflows/main.yml to build the image on PRs
- https://github.com/PrairieLearn/PrairieLearn/blob/master/.github/workflows/images.yml to build/push the image on commits to master

echuber2 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,140 @@
+# Enhanced code-server workspace template for PrairieLearn.
```
I think is the copypasta from the draft of this I shared around to faculty.

echuber2 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,140 @@
+EXPOSE 8080
+
+
+# RUN /bin/sh -c echo "**** install runtime dependencies ****" && \
+# 	sudo apt-get update && \
+# 	sudo apt-get install -y \
+# 	software-properties-common \
+# 	jq \
+# 	libatomic1 \
+# 	net-tools \
+# 	netcat-traditional 
+    	
+# RUN /bin/sh -c echo "**** install Coretto JDK 11 & 17" && \
+# 	sudo apt-get install -y gnupg && \
+# 	#sudo wget -O- https://apt.corretto.aws/corretto.key | sudo apt-key add - && \
+# 	#sudo add-apt-repository 'deb https://apt.corretto.aws stable main' && \
+# 	wget -O - https://apt.corretto.aws/corretto.key | sudo gpg --dearmor -o /usr/share/keyrings/corretto-keyring.gpg && \
+# 	echo "deb [signed-by=/usr/share/keyrings/corretto-keyring.gpg] https://apt.corretto.aws stable main" | sudo tee /etc/apt/sources.list.d/corretto.list && \
+# 	sudo apt-get update && \
+# 	sudo apt-get install -y -11-amazon-corretto-jdk && \
+# 	sudo apt-get install -y java-17-amazon-corretto-jdk  && \
+# 	echo "**** clean up ****" && \
+# 	sudo apt-get clean && \
+# 	sudo rm -rf \
+# 	/config/* \
+# 	/tmp/* \
+# 	/var/lib/apt/lists/* \
+# 	/var/tmp/*
```
I think this is Java-specific stuff from @firasm

firasm replied:
This is all commented out and should have been removed 🙃- sorry!

echuber2 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,140 @@
+
+FROM --platform=linux/amd64 codercom/code-server:4.20.1-bookworm
+
+# On PL, we want to standardize on using 1001:1001 for the user. We change the
+# "coder" account's UID and GID here so that fixuid has no work to do later.
+# This speeds up container loading time drastically and avoids timeouts.
+USER root
+RUN export OLD_UID=$(id -u coder) && \
+	export OLD_GID=$(id -g coder) && \
+	export NEW_UID=1001 && \
+	export NEW_GID=1001 && \
+	groupmod -g 1001 coder && \
+	usermod -u 1001 -g 1001 coder && \
+	find /home -user $OLD_UID -execdir chown -h $NEW_UID {} + && \
+	find /home -group $OLD_GID -execdir chgrp -h $NEW_GID {} + && \
+	unset OLD_UID OLD_GID NEW_UID NEW_GID
```
You did need it previously but weren't doing it, which is why I made the hacked template and began sharing it with staff directly (225 and others are using it). It's needed because the image assumes running as 1000:1000 but PL runs as 1001:1001. The code-server image can try to chown a bunch of stuff at runtime to band-aid the issue, but that is so slow that it times out the startup (225 had this problem specifically).

nwalters512 replied:
That's helpful context! I'm happy to include it if it's necessary, just wanted to make sure I understand why it is.

Tangent: does Jupyterlab require (or would it benefit from) similar treatment?

echuber2 replied:
Jupyter may indeed benefit from similar treatment, although I think they also use something to automatically chown at runtime. So making this change ahead of time is mainly a startup latency benefit.

echuber2 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,140 @@
+COPY --chmod=0444 workspaceTemplate/.vscode /home/coder/project/.vscode
+# COPY --chmod=0444 workspaceTemplate/.lib /home/coder/project/.lib
+
+ENTRYPOINT ["/usr/bin/env", "sh", "/usr/bin/entrypoint.sh", "--bind-addr", "0.0.0.0:8080", "."]
```
A lot of stuff from the hacked entrypoint script can be moved here and we can do away with the hacked entrypoint script, as long as we keed fixuid from running again. I had this in my other draft but I'll try to port it here when I can this week.

echuber2 commented on workspaces/vscode-python/keybindings.json:
```diff
@@ -0,0 +1,2 @@
+// Place your key bindings in this file to override the defaults
+// Write to ./local/share/code-server/User/keybindings.json
```
It looks like Firas added this for other courses to fork and customize the image. I was going to suggest something similar.

firasm commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,140 @@
+		# Install only one extension per line.
+		--install-extension ms-python.python \
+		--install-extension njpwerner.autodocstring \
+		# --install-extension ms-python.debugpy \ # Cannot find the debugger.
```
@Victorsss-Orz how did you find the Default VS Code debugger? 

There is another [Python Debugger](https://marketplace.visualstudio.com/items?itemName=ms-python.debugpy) that Microsoft has, but I don't think it's available in the "Open Source" extension library so we would need to add the .vsix file if we want to use it.

Otherwise, if you think the default VS Code debugger is fine, maybe we leave that one there as-is.

Victorsss-Orz replied:
I've never used other debuggers than the one included in VSCode. The current one supports toggling breakpoints and seeing the variable values between breakpoints. You can just use the "run" tab at the left of the workspace, and choose to run with or without debugging.

firasm replied:
Okay sounds good. You can remove that contented line then

nwalters512 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,140 @@
+ && PATH="/home/coder/miniconda/bin:$PATH" \
+ && /home/coder/miniconda/condabin/conda init bash \
+ && conda config --add channels conda-forge \
+ && conda install ipython \
+ && conda install pandas \ 
+ && conda install seaborn \
+ && conda install pillow
```
These should be pinned to specific versions and put in a separate `requirements.txt` file so that Dependabot can find them. You should also ensure that you add the resulting `requirements.txt` file to our Dependabot config: https://github.com/PrairieLearn/PrairieLearn/blob/master/.github/dependabot.yml

I'd encourage you to match the dependencies and versions included in `prairielearn/grader-python` and `prairielearn/workspace-jupyterlb-python`:

- https://github.com/PrairieLearn/PrairieLearn/blob/033fb69bf120d5747334ee12116e973de0012873/graders/python/requirements.txt
- https://github.com/PrairieLearn/PrairieLearn/blob/033fb69bf120d5747334ee12116e973de0012873/workspaces/jupyterlab-python/requirements.txt

Victorsss-Orz replied:
I don't think PL workspaces uses conda right? Maybe I will remove conda here

nwalters512 commented:
@Victorsss-Orz do you intend to continue work on this PR? If not, we'll close it to keep our list of open PRs focused.

Victorsss-Orz commented:
> @Victorsss-Orz do you intend to continue work on this PR? If not, we'll close it to keep our list of open PRs focused.
Yes I still intend to work on this. I've been caught up in other projects right now, but should have time to improve the code in a week or two.

nwalters512 commented:
@Victorsss-Orz thanks for this! I suspect the issue with `report-image-sizes` is that it doesn't handle newly-created images well. This is on me to fix; I'll let you know when the fix is on `master` so you can pull it in.

Victorsss-Orz commented:
I think the current version is good enough and ready for review. Do people want to discuss what vscode extensions we want to have by default?

nwalters512 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,109 @@
+ && conda config --add channels conda-forge \
+ && conda run -n base pip install -r /requirements.txt
+
+# Cleaning up the cache in the same step when we install is important if you
+# want to minimize the image size.
```
This comment doesn't appear to have any relation to the following line?

echuber2 replied:
This comment refers to the `rm -rf` command that's part of the same `RUN` stanza.

nwalters512 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,109 @@
+# avoid running fixuid if the user is root (for local testing in PL). We also
+# disable auth as PL handles it.
+USER root
+RUN rm -f /run/fixuid.ran && \
```
What is this file that's being removed?

firasm replied:
This one is for @echuber2 I think

echuber2 replied:
If the file exists, it inhibits fixuid from running again. Removing it allows fixuid to run again, which used to be needed in some testing scenarios (although it is slow). But I had a different solution more recently that I'll try to commit to clean this up.

echuber2 replied:
I made a commit that handles this differently by ensuring that the file _exists_ (always inhibit), rather than removing it (to allow the possibility for it to run again). The case where someone might want fixuid to run again is probably not something to worry about. Preventing it from running at all will prevent a local 0:0 session from chowning things to 0:0 or starting slow.

nwalters512 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,109 @@
+# TODO: Maybe customize the preferences to set the theme to auto-adjust to the
+# browser, turn off more telemetry, etc.
+
+# Please read the note here carefully to avoid wiping out what you installed
+# in ~/.local:
```
What was installed in `~/.local`?

firasm replied:
To answer the question, nothing yet, but in `~/.local` is where one would put VSCode workspace settings to enable/disable certain features of VS Code.

In this image, it was decided not to include this (since it's not a template), this is a remnant from @echuber2's comment.

echuber2 replied:
My main point there was to warn instructors about setting the workspace config "home" directory to be `/home/coder` itself, since that name-hides anything that was originally in `/home/coder` as per the image's design.

nwalters512 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,109 @@
+# Please read the note here carefully to avoid wiping out what you installed
+# in ~/.local:
+
+# EDITOR_FOCUS_DIR should be set to the directory you want the editor to start
```
Is this meant to be used by folks building custom images with this one as a base? Even then, I'm 95% sure that once can't retroactively change the based image based on environment variables provided when building an image _from_ that base image, if that makes sense.

Maybe clarify how one is meant to use this environment variable?

echuber2 replied:
This environment variable is only used in the Dockerfile itself in the lines below that, as it ensures the directory exists with mkdir and then sets the working directory to it with WORKDIR. VS Code starts up with the working directory being the focused directory for project work. Realistically, I would have suggested setting this variable to `/home/coder/workspace` and also configuring the PL question with `/home/coder/workspace` as the workspaceOptions "home".

echuber2 replied:
I elided EDITOR_FOCUS_DIR and cut down the comment, also added README.md with a little more explanation (but for now, I won't bother putting too much info in there).

I did want to still mention that people should set `workspaceOptions` home to be `/home/coder/workspace` **and not** `/home/coder`, which they might think is the natural "home directory," but would shadow the settings.

nwalters512 replied:
I think this is an improvement, thanks.

nwalters512 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,109 @@
+		# auto-fix indentation for multiline contexts 
+		--install-extension KevinRose.vsc-python-indent \
+		# visualize indentation with color 
+        --install-extension oderwat.indent-rainbow \
```
Does this need to have extra indentation?

```suggestion
    --install-extension oderwat.indent-rainbow \
```

echuber2 replied:
(Same comment as below,) Removed the extra indentation and the unnecessary () subshell in a commit.

nwalters512 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,109 @@
+		# linting for markdown files
+		--install-extension DavidAnson.vscode-markdownlint \
+		# format tables in markdown files
+        --install-extension fcrespo82.markdown-table-formatter \
```
Does this need to have extra indentation?

```suggestion
    --install-extension fcrespo82.markdown-table-formatter \
```

echuber2 replied:
Removed the extra indentation and the unnecessary () subshell in a commit.

nwalters512 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,109 @@
+USER root
+RUN rm -f /run/fixuid.ran && \
+	echo "" > /usr/bin/entrypoint.sh && chmod 0755 /usr/bin/entrypoint.sh && \
+	/usr/bin/env echo -e '#!/usr/bin/env bash \n' \
+	'set -eu \n' \
+	'# We do this first to ensure sudo works below when renaming the user. \n' \
+	'# Otherwise the current container UID may not exist in the passwd database. \n' \
+	'[ $(id -u) -ne 0 ] && eval "$(fixuid -q)" \n' \
+	' \n' \
+	'if [ "${DOCKER_USER-}" ]; then \n' \
+	'  USER="$DOCKER_USER" \n' \
+	'  if [ "$DOCKER_USER" != "$(whoami)" ]; then \n' \
+	'    echo "$DOCKER_USER ALL=(ALL) NOPASSWD:ALL" | sudo tee -a /etc/sudoers.d/nopasswd > /dev/null \n' \
+	'    # Unfortunately we cannot change $HOME as we cannot move any bind mounts \n' \
+	'    # nor can we bind mount $HOME into a new home as that requires a privileged container. \n' \
+	'    sudo usermod --login "$DOCKER_USER" coder \n' \
+	'    sudo groupmod -n "$DOCKER_USER" coder \n' \
+	' \n' \
+	'    sudo sed -i "/coder/d" /etc/sudoers.d/nopasswd \n' \
+	'  fi \n' \
+	'else \n' \
+	'  sudo rm /etc/sudoers.d/nopasswd \n' \	
+	'fi \n' \
+	'\n' \
+	'dumb-init /usr/bin/code-server --auth none "$@" \n' >> /usr/bin/entrypoint.sh
```
Consider moving this to an external `.sh` file so that this is easier to read and review.

echuber2 replied:
We can actually get rid of this entirely. I'll try to commit an improvement.

echuber2 replied:
Removed this block in a simplifying commit below.

nwalters512 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,109 @@
+	) && \
+	rm -rf /home/coder/.local/share/code-server/CachedExtensionVSIXs
+
+# Prepare the entrypoints. The entrypoint.sh part is based on what code-server
+# originally had, so it may need to be updated when code-server is updated. We
```
Can you link to the original file to make future maintenance easier? https://github.com/coder/code-server/blob/main/ci/release-image/entrypoint.sh

echuber2 replied:
Removed this block in a simplifying commit below. (The original version of the script is still in the base image.)

nwalters512 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,109 @@
+
+EXPOSE 8080
+
+# TODO: Maybe customize the preferences to set the theme to auto-adjust to the
+# browser, turn off more telemetry, etc.
```
Do you intend to resolve this TODO before merging? At least regarding telemetry, it seems like that's already quite disabled.

firasm replied:
This comment can probably just be deleted @Victorsss-Orz - @echuber2 is already tracking this in his future improvement of this.

echuber2 replied:
I'm going to commit a few minor changes to these comments. It's worth noting that, iirc, code-server has an extra option to turn off _their own telemetry_ in addition to the internal settings that disable VS Code's telemetry.

echuber2 replied:
Removed this comment in a commit below.

nwalters512 commented on workspaces/vscode-python/coder.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "query": {
+    "folder": "/home/coder/project"
```
What does this do?

echuber2 replied:
This wasn't part of my version. It looks like this causes code-server to focus a different project directory on startup. However, it's probably not needed, since the default is to use the current working directory (like what I had expected the EDITOR_FOCUS_DIR choice to use). Would vote to remove.

echuber2 replied:
I removed this in a commit below.

nwalters512 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,109 @@
+# up in. This is not necessarily the same as the "home" setting in the
+# question's info.json file. The "home" setting determines the mount point for
+# the persistent cloud storage, which will hide any contents your image
+# originally had at the same path. You might want to set both EDITOR_FOCUS_DIR
+# and "home" to a deeper directory than /home/coder if you want to keep the
+# default home contents from the workspace image (~/.local, etc.). For
+# example, using /home/coder/question will copy the question's workspace
+# folder contents into an empty mount at /home/coder/question and save it for
+# the student, while always reusing the initial contents of /home/coder that
+# you prepared in the image. (However, if students try to customize their
```
This talks about _maybe_ setting `EDITOR_FOCUS_DIR` to something inside `/home/coder`, but below it's already set to `/home/coder/project`. Is this intentional? This feels like it might cause projects where the default of `home` is `/home/coder`; starter files would be copied to e.g. `/home/coder/fibonacci.py`, but the editor would focus on `/home/coder/project` (which would not in fact exist on the mounted-in volume). Or does this new version now require that one sets `"home": "/home/coder/project"`?

echuber2 replied:
Thinking about the "preserving student settings" issue, I think it's not worth setting the question home to `/home/coder` just for that, because the settings will get reset when moving between questions or question instances anyway, and the name-hiding issue of intentional image contents in `/home/coder` might cause problems later on. To preserve student settings, it would be cool if PL had a feature where they could provide a json for that in their site-wide user settings, and have it auto-injected into a workspace. But there would be other compatibility/security issues with that, maybe.

echuber2 replied:
Also, I made a commit below that does use `/home/coder/workspace` instead of `/home/coder/project`.

nwalters512 replied:
> To preserve student settings, it would be cool if PL had a feature where they could provide a json for that in their site-wide user settings, and have it auto-injected into a workspace. But there would be other compatibility/security issues with that, maybe.

I've thought about this, a la the way GitHub Codespaces supports a `dotfiles` repository. The huge obvious problem for this is that this would provide a means to smuggle information into (and possible out of) testing facilities.

I think `/home/coder/workspace` is a reasonable default.

nwalters512 reviewed (commented):
This needs a `README.md` at `workspaces/vscode-python` and an example question.

echuber2 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,81 @@
+
+ENTRYPOINT ["/usr/bin/entrypoint.sh", "--auth", "none", \
+    "--disable-update-check", "--disable-telemetry", \
+    "--bind-addr", "0.0.0.0:8080", "."]
```
Oops, fixed.

echuber2 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,81 @@
+RUN mkdir -p /run /var/run && \
+    touch /run/fixuid.ran /var/run/fixuid.ran
+
+EXPOSE 8080
```
I elided EDITOR_FOCUS_DIR and cut down the comment, also added README.md with a little more explanation (but for now, I won't bother putting too much info in there).

echuber2 replied:
I did want to still mention that people should set `workspaceOptions` home to be `/home/coder/workspace` **and not** `/home/coder`, which they might think is the natural "home directory," but would shadow the settings.

echuber2 replied:
Left the above comments in the wrong place. :man_facepalming: Moving them...

echuber2 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,81 @@
+ && PATH="/home/coder/miniconda/bin:$PATH" \
+ && /home/coder/miniconda/condabin/conda init bash \
+ && conda config --add channels conda-forge \
+ && conda run -n base pip install -r /requirements.txt
```
I added this in a commit below for now, but @Victorsss-Orz might have a reason why not to put it there. The conda stuff is his addition.

echuber2 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,81 @@
+    --install-extension njpwerner.autodocstring \
+    # auto-fix indentation for multiline contexts
+    --install-extension KevinRose.vsc-python-indent \
+    # visualize indentation with color
+    --install-extension oderwat.indent-rainbow \
```
The choice of extensions here are all @Victorsss-Orz so I can't comment on it. Personally, I'd leave it for students to add that themselves if they want it (and if networking is enabled).

firasm replied:
This was me. This can be removed from this image!

echuber2 replied:
Thanks @firasm, I removed that one. What about the others? Do you think they're all suitable?

nwalters512 commented:
@jonatanschroeder this was discussed on Slack and IIRC the consensus was that we'd wait until after this PR was merged and proven out to make that change. I agree that such a base image would be very useful both for first-party and third-party usage.

jonatanschroeder commented:
Sounds good. I'm saying this as a potential user of this image in Python and C, so I'll keep an eye on it.

echuber2 commented:
I see Victor and Firas preferred the Debian upstream image, but I think that code-server offers an Ubuntu one too, which was what I was using. Is there any downside to choosing one or the other, for most courses?

Victorsss-Orz commented:
> I see Victor and Firas preferred the Debian upstream image, but I think that code-server offers an Ubuntu one too, which was what I was using. Is there any downside to choosing one or the other, for most courses?

I didn't bother to change to another upstream image because the current one is already working so there seems no need to change it (yet).

nwalters512 commented:
It actually might be nice to standardize on Ubuntu. We use it as the base for `graders/c`, `graders/java`, and `workspaces/desktop`. This could allow someone to easily build customized grader/workspace images without having to learn how to deal with 2+ distros. Thoughts?

firasm commented:
That sounds reasonable to me! I just picked one and went with it when I developed it.

jonatanschroeder commented:
> It actually might be nice to standardize on Ubuntu. We use it as the base for `graders/c`, `graders/java`, and `workspaces/desktop`. This could allow someone to easily build customized grader/workspace images without having to learn how to deal with 2+ distros. Thoughts?

workspaces/xtermjs uses Debian, should it switch?

echuber2 commented:
Hm, I made some more commits to the branch (switching to Ubuntu, etc.) but the commits aren't showing up in the PR itself here. I'm going to try switching to a draft and back. Also note: https://github.com/orgs/community/discussions/59774

echuber2 commented:
Switching to draft and back didn't work, so I'm not sure what to try. Maybe after a while GH will wake back up.

jonatanschroeder commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,84 @@
+    rm -rf /var/lib/apt/lists/* && \
+    find /tmp -not -path /tmp -delete
+
+# Install Python and required packages via miniconda
+USER coder
+COPY requirements.txt /
+RUN export arch=`uname -m` \
+ && curl -sfLO "https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-${arch}.sh" \
+ && chmod +x "Miniconda3-latest-Linux-${arch}.sh" \
+ && ./"Miniconda3-latest-Linux-${arch}.sh" -b -p /home/coder/miniconda \
+ && PATH="/home/coder/miniconda/bin:$PATH" \
+ && /home/coder/miniconda/condabin/conda init bash \
+ && conda config --add channels conda-forge \
+ && conda run -n base pip install -r /requirements.txt \
+ && rm "Miniconda3-latest-Linux-${arch}.sh" \
+ && conda clean --all
+
+# Install VS Code extensions and clear the extension cache to reduce image size.
+USER coder
+RUN code-server --disable-telemetry --force \
+    # vscode support for python, including debugger
+    --install-extension ms-python.python \
+    # auto-generate templates for function docstrings
+    --install-extension njpwerner.autodocstring \
+    # auto-fix indentation for multiline contexts
+    --install-extension KevinRose.vsc-python-indent \
+    # visualize csv files in vscode window
+    --install-extension mechatroner.rainbow-csv \
+    # linting for markdown files
+    --install-extension DavidAnson.vscode-markdownlint \
+    # format tables in markdown files
+    --install-extension fcrespo82.markdown-table-formatter \
+    && rm -rf /home/coder/.local/share/code-server/CachedExtensionVSIXs
```
Keeping in mind the possibility that a future PR would split the base and Python-specific content into separate images, would it make sense to be prepared for that and move all Python-specific commands to the bottom of the Dockerfile?

nwalters512 replied:
I'm fine keeping the order largely how it is, though you might consider moving extension installation above the `requirements.txt` stuff for better layer caching/reuse?

echuber2 replied:
I thought there was some reason why the Miniconda Python stuff needed to be installed before the extension installs would complete successfully, but I haven't tested that. (I thought one of the others mentioned this to me, but I could be misremembering.)

nwalters512 replied:
I think if that's the case, we should document why the order matters, I often find it helpful to leave comments about when ordering matters implicitly.

echuber2 replied:
@firasm I could try to build this with the order inverted, but from your memory, does the installation of the Python extensions require the Conda installs to complete first? (If I try it and it _seems_ to work, I'm not sure if it will actually be fully working.) Or we can add a comment to document this as Nathan says.

firasm replied:
The Python extension from Microsoft clearly states that a supported version of python should be installed first so we should just follow that guidance and add a comment 

![Screenshot 2024-05-30 at 1 08 29 PM](https://github.com/PrairieLearn/PrairieLearn/assets/2507459/d291388e-0758-4e0d-b024-ff144a8fe715)

Of course, nothing to say it WON'T work if the order is inverted, but why play with fire 🙃

echuber2 replied:
I put comments in the Dockerfile to this effect now.

echuber2 replied:
It's worth noting here (at least) that the extension gets installed from the Open VSX marketplace, not from the MS one, but the extension is in fact a republished version of Microsoft's extension. Open VSX did that within the terms of the license, and MS also commented on their GitHub repo that it was okay to do so with a custom-built vsix.

echuber2 commented:
Would it be okay to add the setting `"window.menuBarVisibility": "visible"`? This nests a traditional menubar in the top bar instead of putting a hamburger button. It doesn't take up more screen space because the top bar is there anyway, and it seemed more discoverable or more accessible to me somehow. (I think it's especially useful if people aren't sure how to pop the terminal, since the Terminal menu is readily visible.)

Here's a screenshot of the default "compact" setting on the left (after clicking the hamburger button to display it), and the "visible" setting on the right:

![vscode-menubar-comp](https://github.com/PrairieLearn/PrairieLearn/assets/5503940/c11d42b0-9a91-4fd3-89a8-b4c760f12d0e)

firasm commented:
> Would it be okay to add the setting `"window.menuBarVisibility": "visible"`?

Those are good and valid reasons to include this setting from my side, especially because the bar is already there!

echuber2 commented:
So I added the setting for the menu bar and also included a `tasks.json` file that ensures the terminal panel will open automatically at startup time. I think this detail was requested by one of our professors last year.

nwalters512 commented:
@echuber2 @Victorsss-Orz what's the status of this PR, is it ready for another review? Is there anything else we want to add here?

Victorsss-Orz commented:
Actually I think I still need to create an example question using this workspace? Or can we start the review without this

echuber2 commented:
Apart from the example question issue, I wasn't going to change anything further right now.

Victorsss-Orz commented:
The end of semester is starting to get crazy. I'll be sure to do it in early May. Is that ok?

echuber2 commented:
Since there was already an example question referencing the old prototype image that's technically been unsupported, I went ahead and just revised that question to serve as an example.

While I was doing that, I realized we really did need the gosu helper to make sure to step down from 0 to 1001 in local dev mode, so I added that now rather than make a followup PR. (In the past, I included such a script with the RStudio workspace.)

echuber2 commented:
@Victorsss-Orz When you have time to look at this, I think the revised demo question may be sufficient now as a minimal placeholder, but maybe you'd like to add another question.

Victorsss-Orz commented:
> @Victorsss-Orz When you have time to look at this, I think the revised demo question may be sufficient now as a minimal placeholder, but maybe you'd like to add another question.

Maybe we can wait and get some more reviews before I try to add anything else?

Victorsss-Orz commented:
> @Victorsss-Orz When you have time to look at this, I think the revised demo question may be sufficient now as a minimal placeholder, but maybe you'd like to add another question.

Maybe we can wait and get some more reviews before I try to add anything else?

echuber2 commented:
Sure, @nwalters512 if it looks okay for now, can you let @Victorsss-Orz know if it's okay to add more? Or maybe the PR can be finalized now and followup features can be separate.

nwalters512 commented on workspaces/vscode-python/requirements.txt:
We should double-check that the versions here match those in the latest `prairielearn/grader-python` and `prairielearn/workspace-jupyterlab-python` images.

nwalters512 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,89 @@
+    --install-extension KevinRose.vsc-python-indent \
+    # visualize csv files in vscode window
+    --install-extension mechatroner.rainbow-csv \
+    # linting for markdown files
+    --install-extension DavidAnson.vscode-markdownlint \
```
I used this locally for a little bit but eventually turned it off because of a bad signal to noise ratio. IMO we should not include this.

firasm replied:
I don't remember adding this, can be removed !

nwalters512 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,89 @@
+RUN code-server --disable-telemetry --force \
+    # vscode support for python, including debugger
+    --install-extension ms-python.python \
+    # auto-generate templates for function docstrings
+    --install-extension njpwerner.autodocstring \
```
This feels oddly opinionated; is it necessary to include this?

echuber2 replied:
I'll just say that these specific extensions were chosen by the others, not me. So, they'll have to comment about that.

nwalters512 replied:
@firasm I think this all originated with you. Can you argue for the inclusion of each of these extensions?

firasm replied:
Happy to have this, and any other opinionated extensions removed!

Victorsss-Orz commented:
@echuber2 I just removed the two vscode extensions you were discussing about. Currently the network environment I'm in doesn't allow me to test the docker installation so I haven't modified the requirements.txt yet. Do you have the time to update the requirements.txt to be the same as the grader image and see the it can be built? Earlier when I was working on the extensions, I couldn't get pygraphviz to work so I just removed it from the requirements. Maybe the building can be successful now? I'm not sure

echuber2 commented:
I'm making some progress on sorting out the conda problems and I'll make a commit soon. Since you switched to miniforge on the grader, I tried using it here too, and using mamba instead of conda was both faster and avoided an error conda was having with some dynamic library resolution. However there's enough complexity that we might want to move the conda install script out to a separate file, like with the grader.

echuber2 commented:
It doesn't look like the size was reported by CI, but it's roughly 2.25 GB.

echuber2 commented:
I also tried installing everything in requirements.txt that can be found in conda-forge, which mamba tries to use exclusively, and only using pip for the few that remain (matplotlib==3.9.0, openpyxl==3.1.3, typing-extensions==4.12.0). That worked too, but the resulting image would be 2.84 GB. I'm not sure which would be more performant, but someone else can decide about that.

firasm commented:
Hi @echuber2 - are you finished adding stuff? Can we try to get this merged in? A few people have asked about workspaces and it would be good to point them to this example.

echuber2 commented:
@firasm Yes, I've been ready to have it merged for months, basically. I think I fixed the new issue that Victor raised, but we still need Nathan's final approval. Can you test it too? Today is a holiday in the US so it may be a slow week.

Victorsss-Orz commented:
I can also try to test it locally - Should I build the docker image first and have it named to prairielearn/workspace-vscode-python before trying to access the example question?

echuber2 commented:
Yes, name the image exactly how the question expects it to be. Here's how I test locally: https://github.com/PrairieLearn/PrairieLearn/discussions/8897

The `"workspacePullImagesFromDockerHub": false` makes sure that it uses your prebuilt local copy of the image without pulling and overwriting it. However in this case it might not matter, because I think the image hasn't been published at all to Docker Hub yet.

echuber2 commented:
The last two commits are just comments. I noticed that the most recent Ubuntu images have a pre-installed `ubuntu` user 1000:1000, so as a result, *the coder user is already 1001:1001*. The uid adjustment code is still good to keep as error prevention in case this changes down the line, or if PL decides to go with Debian (or other) base images instead.

Also worth noting is that the newer VS Code upstream reportedly had some issues on macOS, so it may be best to skip a version of code-server while the fix propagates.

nwalters512 commented:
> Also worth noting is that the newer VS Code upstream reportedly had some issues on macOS, so it may be best to skip a version of code-server while the fix propagates.

@echuber2 can you link to the relevant issue/fix so we can determine what version we can/can't use?

nwalters512 commented on exampleCourse/questions/demo/workspace/vscode-python/question.html:
```diff
@@ -1,7 +1,14 @@
 
-  <p><strong>Note to course developers:</strong> The code-server image is pinned to version 3.4.1 because 3.5.0 is currently affected by an intermittent "could not reconnect" bug. See <a href="https://github.com/cdr/code-server/issues/2144">cdr/code-server#2144</a></p>
+  <p class="alert alert-info" role="alert">
+    <strong>Note to course developers:</strong> We use an open-source, web-based version of VS Code
+    (<a href="https://github.com/coder/code-server" target="_blank">code-server</a>).
+    To comply with Microsoft's license terms, code-server installs extensions from the independent
+    marketplace <a href="https://open-vsx.org" target="_blank">Open VSX</a> instead of using Microsoft's
+    official VS Code marketplace. Many extensions are available from Open VSX by the same names,
+    but some extensions might not be compatible.
```
This feels like information that belongs in the image documentation/readme, not an example question. Is there are good reason this is here? If not, I'll move it.

echuber2 replied:
OK, where would it be best to move it?

echuber2 replied:
For now, I moved this into the `README.md` for the workspace itself.

nwalters512 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,105 @@
+# noble is the codename for Ubuntu 24.04 LTS
+FROM codercom/code-server:4.90.3-noble
+# Skip code-server:4.91.0-noble and delay updating beyond this until
+# code-server catches up to vscode 1.91.1 because of possible issues on
+# macOS. See: https://github.com/microsoft/vscode/issues/219893
```
If this issue is unique to macOS, how could it possibly affect PL where the Docker image will always be running on Linux (either natively on Linux machines, or in the Linux VM locally on macOS/Windows)?

echuber2 replied:
I'd thought the issue was specific to _Safari_ but I may have misunderstood that when I viewed the issue discussions.

echuber2 replied:
The promised version with the fix has already propagated to code-server to I'll try just updating the base image now, and then we can remove the warning.

nwalters512 commented on workspaces/vscode-python/Dockerfile:
```diff
@@ -0,0 +1,105 @@
+ && . "/home/coder/conda/etc/profile.d/conda.sh" \
+ && . "/home/coder/conda/etc/profile.d/mamba.sh" \
+ && mamba activate \
+ # Mamba can install pygraphviz without compile errors if we do so first, separately from the rest of the requirements:
+ && mamba install --yes --quiet $(grep pygraphviz /requirements.txt | head -n 1) \
```
This feels very fragile and isn't something we've faced with other Python images. Why is this? Have you reported this upstream to the Mamba folks?

Victorsss-Orz replied:
I took a look at the python grader image and the jupyterlab image. Both install some packages that are required to build the image, but even with the same packages installed in the vscode image the pygraphviz wheel building still fails. I have not been able to identify the source of error yet. 

With that being said, I don't mind using this extra tweaking for pygraphviz at the moment so that the image can be publicly available. I also haven't used pygraphviz so I can't tell whether it is a good idea to just remove the package.

echuber2 replied:
@Victorsss-Orz Is the current version in PR here failing to build for you too? (At least, it was building for me, but maybe it's not deterministic.) @firasm any thoughts on how to deal with pygraphviz here, or remove it instead?

Victorsss-Orz replied:
I can build the image with the current version. 
I've tried removing this line for pygraphviz and install some of the dependencies directly but it failed. I'm fine with using this hack for pygraphviz only, or if people want that removed it would also be ok.

firasm replied:
I'd prefer removing Pygraphviz then, document it's missing, and create an issue that captures the reason why so it can be handled in a future PR. 

I think Nathan is right, adding these tweaks to the initial image is asking for trouble later and people will copy/paste it and it'll persist forever.

echuber2 replied:
OK, we can do that, but since (py)graphviz is in the grader image, there will be some discrepancy. I think that one or more of the Python packages must have had an inaccurate list of dependencies, so where the grader installs the system package for graphviz first instead, it's basically doing the same workaround to sidestep the issue.

firasm replied:
Yes I think we accept that the first version won't make 100% of everyone 100% happy but perfect is the enemy of good!

@Victorsss-Orz or @echuber2 would either of you like to tackle this? I think that's the last remaining obstacle.

Victorsss-Orz replied:
Yes I can try to fix this and do the last round of testing by the end of tomorrow (if not today)

echuber2 replied:
> Yes I think we accept that the first version won't make 100% of everyone 100% happy but perfect is the enemy of good!

To be honest I don't think I'm the one insisting on perfection. I'm trying to understand the changing criteria to get this merged. Anything is fine with me.

daviddalpiaz replied:
Just jumping into to mention that I'm not dependent on this for the start of the semester. (Don't want y'all to rush because of my comment below.) I ended up making my own. I'll be curious to see what the official image looks like in the end though.

(I mostly just skipped `pygraphviz` and installed other things needed for my course.)

nwalters512 replied:
It's not my intention to be constantly changing requirements, but it seems like every time I come back to look at this PR, something big has changed (new Ubuntu! Mamba! version pinning!) that I'm going to respond to. Both implementation and review here seem to be a moving target.

I want to get this merged too, clearly this image is in demand. I'll try to spend a little time here tomorrow to see if I can't improve things a bit on my own, while also keeping in mind that perfect is indeed the enemy of good. Thanks for your patience and work, everyone.

Victorsss-Orz replied:
I have a local version without pygraphviz and the rest works just fine. Should I wait until you try things tomorrow to push this?

nwalters512 replied:
I poked at this a bit. If I also installed `build-essentials graphviz graphviz-dev` at the OS level, pygraphviz was able to build successfully, but at the cost of a larger image size (2.41GB vs 2.23GB when it's installed with Mamba).

I was hopeful that we could use `mamba` to install _all_ packages to make things less idiosyncratic, but some of the package versions we use don't seem to be available via Conda:

```
13.25 Pinned packages:
13.25   - python 3.10.*
13.25 
13.25 
13.25 Could not solve for environment specs
13.25 The following packages are incompatible
13.25 ├─ matplotlib 3.9.0  does not exist (perhaps a typo or a missing channel);
13.25 ├─ openpyxl 3.1.3  does not exist (perhaps a typo or a missing channel);
13.25 └─ typing-extensions 4.12.0  does not exist (perhaps a typo or a missing channel).
```

Given this, I think doing the mamba/pip split is fine. I updated the documentation in 79002ac to better explain why we're doing things like this (in the end, it's not _just_ to "install without compiler errors").

nwalters512 replied:
@Victorsss-Orz sorry, I missed your note before I pushed and commented. Let's keep `pygraphviz` installed for parity with our other Python images. I'm a lot more comfortable with the state of things now that I better understand how/why this works.

echuber2 replied:
I think I had tried throwing Mamba at the entire requirements file and eventually settled on the split when I saw the same problem. Sorry I didn't take more detailed notes about it to justify the split.

daviddalpiaz commented:
Any chance of this being merged before the start of the UIUC semester? (Reading the above makes it seem like this is close, so I wanted to check before I try to hack something together myself.)

nwalters512 reviewed (approved):
I'm happy where this has ended up. Victor/Eric, I'll let you give a final sign-off on my tweaks from yesterday, and then we can merge this 🎉🚀

Victorsss-Orz commented:
I'm glad the package problems are (mostly) sorted out. I think we can make this available for people to use.
